puma/hlplots/results.py

```diff
@@ -497,15 +501,116 @@ def plot_var_perf(  # pylint: disable=too-many-locals
         if h_line:
             plot_sig_eff.draw_hline(h_line)
 
-        plot_base = "profile_flat" if kwargs.get("fixed_eff_bin") else "profile_fixed"
-        plot_suffix = f"{self.signal}_eff_{suffix}" if suffix else f"{self.signal}_eff"
-        plot_sig_eff.savefig(self.get_filename(plot_base, plot_suffix))
+        plot_base = (
+            "profile_flat_per_bin"
+            if kwargs.get("flat_eff_bin")
+            else "profile_fixed_cut"
+        )
+        plot_details = f"{self.signal}_eff_vs_{x_var}_"
+        plot_suffix = f"_{suffix}" if suffix else ""
+        plot_sig_eff.savefig(self.get_filename(plot_details + plot_base, plot_suffix))
+        for i, background in enumerate(self.backgrounds):
+            plot_bkg[i].draw()
+            plot_details = f"{background}_rej_vs_{x_var}_"
+            plot_bkg[i].savefig(
+                self.get_filename(plot_details + plot_base, plot_suffix)
+            )
+
+    def plot_flat_rej_var_perf(
+        self,
+        fixed_rejections: dict[Flavour, float],
+        suffix: str = None,
+        xlabel: str = r"$p_{T}$ [GeV]",
+        x_var: str = "pt",
+        h_line: float = None,
+        **kwargs,
+    ):
+        """Plot signal efficiency as a function of a variable, with a fixed enforce
+        background rejection for each bin
+
+
+        Parameters
+        ----------
+        fixed_rejections : dict[Flavour, float]
+            A dictionary of the fixed background rejections for each flavour, eg:
+            fixed_rejections = {cjets : 0.1, ujets : 0.01}
+        suffix : str, optional
+            suffix to add to output file name, by default None
+        xlabel : regexp, optional
+            _description_, by default "$p_{T}$ [GeV]"
+        x_var: str, optional
+            The x axis variable, used for providing details to the plot name,
+            default is 'pt'
+        h_line : float, optional
+            draws a horizonatal line in the signal efficiency plot
+        **kwargs : kwargs
+            key word arguments for `puma.VarVsEff`
+        """
+        assert all(
+            [b.name in fixed_rejections for b in self.backgrounds]
+        ), "Not all backgrounds have a fixed rejection"
+        plot_bkg = []
+        for background in self.backgrounds:
+            modified_second_tag = (
+                f"{self.atlas_second_tag}\nFixed {background.rej_str} ="
+                f" {fixed_rejections[background.name]} per bin"
```

Comment:
I'm wondering if we could modify the second tag inside the `VarVsEffPlot` based on it's arguments. That way the second tag is always clearly corresponding to the plot content, even if the HLAPI is not used. Would that be straightforward?

Reply:
also I would replace the `=` with `of`

Reply:
The only issue with doing it inside the VarVsEffPlot class rather than Results, is that we don't have access to what we're using as a background or signal. So it would be possible to have 'rejection of X per bin', but not 'c-jet rejection of X per bin', which inherently then means we're not including all relevant information in the string.

Reply:
We could just pass the extra info as an additional argument, but maybe that's more trouble than it's worth, happy for the Results class to keep handling this for now in that case

Reply:
So, looking back into this, this is again non trivial. The plots allow each tagger to have individual values for flat/fixed working point/didc cut, as is shown [here](https://github.com/nikitapond/puma/blob/improve_peff_plots/puma/hlplots/results.py#L468-L472).
This I think leaves us with several options:

- Force the user to pass either working cut or disc cut as an argument, so we can set add a relevant 'flat 70% b-efficiency' tag.
- Allow the user to not pass either of these, and use tagger defaults, but if these aren't passed, then just have 'flat varying b-eff' as the tag, and encourage the user to include these values in their model names.
- As above, but automatically include whatever the taggers disc cut/WP is in the tagger name.

I think the last option makes the most sense, so I'll try implement that. If you disagree, either of the other two options are much simpler to implement.

Reply:
wow that is annoying - I can't think there is a great use case for plotting taggers with different working points. So you are suggesting to add the disc cut / WP to the tagger label in the plot legend? I think I prefer to put the "flat per bin blah" strings in the `atlas_second_tag` rather than the tagger label names.

Should we remove those lines you liked to which use the tagger disc cut for these plots?

Reply:
Yeah IMO I agree, I can't think of the use case. @manuguth included these in the original HL API, perhaps we can check if he thinks there's a requirement for them, or if we can do away with them?

Reply:
He left ATLAS now sadly, so might not respond here. I'd say go ahead :)

Reply:
Could it be that this was originally designed to had over the working point (like 77%) but set the cut value to a certain value? This could be used if you want to plot the ttbar 70% WP Cut value for Zprime. Not 100% sure if I got the code right tbh.

Reply:
<img width="954" alt="image" src="https://github.com/umami-hep/puma/assets/63856590/cded5212-89b0-4961-8139-094857f96e88">
'fixed cut'

<img width="1022" alt="image" src="https://github.com/umami-hep/puma/assets/63856590/23144165-9e85-4f6c-97d0-903a4f505d07">
'flat per bin'


<img width="1026" alt="image" src="https://github.com/umami-hep/puma/assets/63856590/d0269de5-e3c7-43c0-b199-f124fbc3d50f">
'flat rejection per bin'